concurrent/src/main/java/io/airlift/concurrent/BoundedExecutor.java

```diff
@@ -52,35 +50,42 @@ public BoundedExecutor(Executor coreExecutor, int maxThreads)
     @Override
     public void execute(Runnable task)
     {
-        checkState(!failed.get(), "BoundedExecutor is in a failed state");
-
         queue.add(task);
 
         int size = queueSize.incrementAndGet();
         if (size <= maxThreads) {
-            // If able to grab a permit (aka size <= maxThreads), then we are short exactly one draining thread
             try {
                 coreExecutor.execute(this::drainQueue);
             }
             catch (Throwable e) {
-                failed.set(true);
-                log.error("BoundedExecutor state corrupted due to underlying executor failure");
+                decrementAndGetQueueSize();
                 throw e;
             }
         }
     }
 
     private void drainQueue()
     {
-        // INVARIANT: queue has at least one task available when this method is called
-        do {
-            try {
-                queue.poll().run();
+        while (true) {
+            Runnable task = queue.poll();
+
+            if (task != null) {
+                try {
+                    task.run();
+                }
+                catch (Throwable e) {
+                    log.error(e, "Task failed");
+                }
             }
-            catch (Throwable e) {
-                log.error(e, "Task failed");
+
+            if ((decrementAndGetQueueSize() == 0) && (task == null)) {
+                return;
```

Comment:
With maxThreads=1 consider

1. thread A calls execute
   a. adds task
   b. queueSize := 1
   c. starts a thread A'
1. thread A' removes the task from the queue and executes it
1. thread A' calls `decrementAndGetQueueSize`, i.e. queueSize := 0, but does not exit because `task` was not null
1. thread B (or A) calls execute 
   a. adds task
   b. queueSize := 1
   c. starts a thread B', because queueSize <= maxThreads

At this point, 2 threads A' and B' are executing within coreExecutor even though maxThreads is 1.

Reply:
(updated)

Reply:
But if we could rearrange the `if` condition to `task == null && decrementAndGetQueueSize() == 0` then threadB' won't be created as `queueSize := 0` only if the `task == null` or maybe we can peek into the queue to check for the next task 
`decrementAndGetQueueSize == 0 && queue.peek() == null`